connectors/src/connectors/notion/temporal/workflows/index.ts

```diff
@@ -222,56 +266,100 @@ export async function notionGarbageCollectionWorkflow({
 
   const promises: Promise<void>[] = [];
 
-  // we go through each result page of the notion search API
-  do {
-    // It's a garbage collection, we want to fetch all pages.
-    const { pageIds, databaseIds, nextCursor } =
-      await getPagesAndDatabasesToSync({
-        connectorId,
-        lastSyncedAt: null,
-        // Only pass non-null cursors.
-        cursors: cursors.filter((c) => c !== null) as string[],
-        excludeUpToDatePages: false,
-        loggerArgs: {
+  async function runSearch(
+    cursorType: "pages" | "databases"
+  ): Promise<{ isComplete: boolean; pageIndex: number }> {
+    let pageIndex = pageIndexes[cursorType];
+    const { last: lastCursor } = cursors[cursorType];
+    if (pageIndex > 0 && !lastCursor) {
+      // We are already done
+      return {
+        isComplete: true,
+        pageIndex,
+      };
+    }
+    // we go through each result page of the notion search API
+    do {
+      // It's a garbage collection, we want to fetch all pages.
+      const { pageIds, databaseIds, nextCursor } =
+        await getPagesAndDatabasesToSync({
+          connectorId,
+          lastSyncedAt: null,
+          // Only pass non-null cursors.
+          cursors: cursors[cursorType],
+          excludeUpToDatePages: false,
+          loggerArgs: {
+            pageIndex,
+            runType: "garbageCollection",
+          },
+          filter: cursorType === "pages" ? "page" : "database",
+        });
+
+      // Update the cursors object to keep both the previous and last cursors.
+      const newPreviousCursor = cursors[cursorType].last;
+      const newLastCursor = nextCursor;
+      cursors[cursorType] = {
+        previous: newPreviousCursor,
+        last: newLastCursor,
+      };
+
+      pageIndex += 1;
+
+      // this function triggers child workflows to process batches of pages and databases.
+      // the worflow that processes databases will itself trigger child workflows to process
+      // batches of child pages.
+      promises.push(
+        performUpserts({
+          connectorId,
+          pageIds,
+          databaseIds,
+          isGarbageCollectionRun: true,
+          runTimestamp,
           pageIndex,
-          runType: "garbageCollection",
-        },
-      });
+          isBatchSync: false,
+          queue: childWorkflowQueue,
+          topLevelWorkflowId,
+          forceResync: false,
+        })
+      );
 
-    // Update the cursors array to keep only the last 2 cursors.
-    cursors = [cursors[1] ?? null, nextCursor];
+      if (pageIndex % 512 === 0) {
```

Comment:
🤔 So we go over max 512 pages?

Reply:
after 512 pages we continue as new (but we continue from where we left off).
Something @Fraggle added to fix GC workflows going over history size limit. And it works well which is how we even know that some GCs run infinitely

Reply:
👍 IMHO worth a comment 😛